chore(docs): port MORE notions (#180)
sapphiredev/website: docs/Guide/application-commands/application-command-registry/registering-chat-input-commands.mdx

title: Registering Chat Input Commands
---

A chat input command can be registered to the Discord API by letting the command registry know about its existence. To

> do so we need to acquire or use the [`registerApplicationCommands`][rac] to have access to the

Reviewer: ```suggestion
do so we need to use the [`registerApplicationCommands`][rac] to have access to the
```

> [`ApplicationCommandRegistry`][acr] to perform operations on it. You can learn how to do this on the page [Acquiring an

Reviewer: ```suggestion
[`ApplicationCommandRegistry`][acr] or acquire one to perform operations on it. You can learn how to do this on the page [Acquiring an
```

> [`ApplicationCommandRegistry`][acr] to perform operations on it. You can learn how to do this on the page [Acquiring an
> Application Command Registry][aaacr].

Reviewer: ```suggestion
[`ApplicationCommandRegistry`][acr] to perform operations on it. You can learn how to do this on the page [`Acquiring an
Application Command Registry`][aaacr].
```

## How to register a chat input command

> The easiest way to register a Chat Input Command is to implement the [`registerApplicationCommands`][rac] method on a

Reviewer: ```suggestion
The easiest way to register a chat input command is to implement the [`registerApplicationCommands`][rac] method on a
```

Reviewer 2: We capatilize it everywhere and I think that's better

[`Command`][command] class. In that method the first parameter is the [`ApplicationCommandRegistry`][acr] (commonly
called `registry` in code) on which you can call [`registerChatInputCommand`][rcic]. This looks like:

> ```ts ts2esm2cjs|{7,11-17}|{7,11-17}

Reviewer: Is the duplicated highlights needed?

Reviewer 2: Yeah I never added copying JS highlights to TS when the latter is omitted.

import { Command } from '@sapphire/framework';

export class SlashCommand extends Command {
  public constructor(context: Command.Context, options: Command.Options) {
    super(context, {
      ...options,
      description: 'Send a Slash Command.'
    });
  }

  public override registerApplicationCommands(registry: Command.Registry) {
    registry.registerChatInputCommand((builder) =>
      builder //
        .setName(this.name)
        .setDescription(this.description)
    );
  }

  public override chatInputRun(interaction: Command.ChatInputInteraction) {
    // ...
  }
}
```

:::info

We use `//` next to `builder //` here to short-circuit Prettier putting everything on the same line. If you do not use

> prettier you can remove this, or if you do not care for it being on the same line then you can remove it also.

Reviewer: ```suggestion
prettier you can remove this, or if you do not care for it being on the same line then you can remove it.
```

:::

In this code snippet we have registered the Chat Input Command in the most simple way referencing the already defined
`name` and `description`. We also use the [`builder`][djs-slash-builder-docs] pattern which means you get verification
of your input before a command gets registered in a faulty way. It will also make sure that you are not providing
unnecessary or incorrect JSON keys because you're not building the JSON yourself. All that said, it is the method we

> recommend everyone to use. If you are experienced with Chat Input Commands you can also opt to write your own raw JSON

Reviewer: ```suggestion
recommend everyone to use. If you are experienced with chat input commands you can also opt to write your own raw JSON
```

object, however this guide will not cover that.

In the code snippet above we passed 1 parameter to the [`registerChatInputCommand`][rcic] method, to build the object
that gets send to Discord for registering the command. The [`registerChatInputCommand`][rcic] however also takes a
second parameter which we will cover next.

### Registering Chat Input Commands with options

> Shown above is how to register a Chat Input Command without options, but in many cases you will actually want to receive

Reviewer: ```suggestion
Shown above is how to register a chat input command without options, but in many cases you will actually want to receive
```

In this example we will add an option that will prompt the user of the command to select another user that they want to
greet. This option is required and the command will not send at all if the option is not provided. This is really useful
as compared to Sapphire [Arguments][arguments] you no longer have to validate whether a required argument was provided
or not.

```ts ts2esm2cjs|{16-21}|{16-21}
import { Command } from '@sapphire/framework';

export class SlashCommand extends Command {
  public constructor(context: Command.Context, options: Command.Options) {
    super(context, {
      ...options,
      description: 'Say hello to a user.'
    });
  }

  public override registerApplicationCommands(registry: Command.Registry) {
    registry.registerChatInputCommand((builder) =>
      builder //
        .setName(this.name)
        .setDescription(this.description)
        .addUserOption((option) =>
          option //
            .setName('user')
            .setDescription('User to say hello to')
            .setRequired(true)
        )
    );
  }

  public override chatInputRun(interaction: Command.ChatInputInteraction) {
    // ...
  }
}
```

> There are many other types of options that you can add to your Chat Input Command, and you can read all about them on

Reviewer: ```suggestion
There are many other types of options that you can add to your chat input command, and you can read all about them on
```

you have configured the Sapphire Logger to only log errors or fatal messages then please check
[`Configuring Log Level Guide`][configure-logging] to learn how to change it.

### `guildIds`

This is an array of guild ids for which the command should be registered. By default it is `undefined`, which means that
the command will be registered for all guilds (i.e. globally). If you want to register the command for a specific guild,
you can pass in the guild id in the command registry.

:::note

When the [`guildIds`][guildids] is an empty array the command will also be registered globally. This is useful for when
you're getting the [`guildIds`][guildids] from an external source, i.e. through a `getGuildIds` function. You can make
that external source default to an empty array which is often easier than `undefined`, especially when it comes from an
external source that using a difference programming language from JavaScript/TypeScript (i.e. an external API or
database).

:::

### `behaviorWhenNotIdentical`

This options tells the registry that how to behave when the command is not identical to the one which is already

> registered on Discord API. By default it's set to [`LOG_TO_CONSOLE`][behavior-log] which means that it will only warn

Reviewer: Update this to OVERWRITE

Reviewer 2: Sure, once you PR that to framework

So in 3 years

you about the data differences found as compared to the data received from the Discord API.

If you want the registry to be overwritten with your own local data, you can set this option to

> [`OVERWRITE`][behavior-overwrite]. Note that configuring [`OVERWRITE`][behavior-overwrite] will also still log the

Reviewer: Half true, it logs the differences at debug level

Reviewer 2: No, this is entirely true. Log levels are completely irrelevant to this sentence. And any log level is "will also still log...".

Also this is not really useful feedback. Either suggest a different wording, or don't say anything lol.

Resolving without change.

change to the console. That is to say, [`OVERWRITE`][behavior-overwrite] includes [`LOG_TO_CONSOLE`][behavior-log]
rather than both being separate code branches.

### `registerCommandIfMissing`

This is an option that defaults to `true`. If set to `false`, the command will not be registered if it is not already

> registered i.e. if not already present in the Discord API.

Reviewer: ```suggestion
registered in the Discord API.
```

:::info

You will likely never have to touch this option, but it can be useful if you are not using Sapphire to register your
commands and instead you're using something external to do so.

:::

## Chat Input Command Registry Methods

:::warning

To register a chat input command, you need to acquire the [`ApplicationCommandRegistry`][acr]. Please see the

> [`ApplicationCommandRegistry`][aaacr] section for more details.

Reviewer: Two sections?

Reviewer 2: Yes, the top one is the docs and the latter is the guide.

application commands automatically for you else consider using `idHints` instead.

> Whenever `addChatInputCommandNames` or `addChatInputCommandIds` method is used and wrong value/id is provided, then it

Reviewer: ```suggestion
Whenever the `addChatInputCommandIds` method is called with an invalid id (we expect a string that can be turned into a BigInt), then it
```

`addChatInputCommandNames` or `addChatInputCommandIds` should only be used when you don't want `Sapphire` to manage the
application commands automatically for you else consider using `idHints` instead.

Whenever the `addChatInputCommandIds` method is called with an invalid id (we expect a string that can be turned into a
BigInt), then it will **ALWAYS** warn in the console. No, this can't be turned off.

:::

## Full Example

> Now that we have covered everything there is to registering a Chat Input Command lets look at how a full example could

Reviewer: ```suggestion
Now that we have covered everything there is to registering a chat input command lets look at how a full example could
```
